packages/transaction-controller/src/helpers/ResimulateHelper.ts

```diff
@@ -1,31 +1,139 @@
 import type { Hex } from '@metamask/utils';
-import { createModuleLogger, remove0x } from '@metamask/utils';
+import { remove0x } from '@metamask/utils';
 import { BN } from 'bn.js';
 import { isEqual } from 'lodash';
 
-import { projectLogger } from '../logger';
+import { createModuleLogger, projectLogger } from '../logger';
+import { TransactionStatus } from '../types';
 import type {
   SimulationBalanceChange,
   SimulationData,
   TransactionMeta,
   TransactionParams,
 } from '../types';
-import { getPercentageChange } from './utils';
+import { getPercentageChange } from '../utils/utils';
 
-const log = createModuleLogger(projectLogger, 'resimulate');
+const log = createModuleLogger(projectLogger, 'resimulate-helper');
 
 export const RESIMULATE_PARAMS = ['to', 'value', 'data'] as const;
 export const BLOCKAID_RESULT_TYPE_MALICIOUS = 'Malicious';
 export const VALUE_COMPARISON_PERCENT_THRESHOLD = 5;
 export const BLOCK_TIME_ADDITIONAL_SECONDS = 60;
+export const RESIMULATE_INTERVAL_MS = 3000;
 
 export type ResimulateResponse = {
   blockTime?: number;
   resimulate: boolean;
 };
 
+export type ResimulateHelperOptions = {
+  getTransactions: () => TransactionMeta[];
+  onStateChange: (listener: () => void) => void;
+  updateSimulationData: (transactionMeta: TransactionMeta) => Promise<void>;
+};
+
+export class ResimulateHelper {
+  // Map of transactionId <=> isActive
+  readonly #activeResimulations: Map<string, boolean> = new Map();
+
+  // Map of transactionId <=> intervalId
+  readonly #intervalIds: Map<string, NodeJS.Timeout> = new Map();
+
+  readonly #getTransactions: () => TransactionMeta[];
+
+  readonly #updateSimulationData: (transactionMeta: TransactionMeta) => Promise<void>;
+
+  constructor({
+    getTransactions,
+    updateSimulationData,
+    onStateChange,
+  }: ResimulateHelperOptions) {
+    this.#getTransactions = getTransactions;
+    this.#updateSimulationData = updateSimulationData;
+
+    onStateChange(() => {
+      const unapprovedTransactions = this.#getUnapprovedTransactions();
+      const unapprovedTransactionIds = new Set(
+        unapprovedTransactions.map((tx) => tx.id),
+      );
+
+      // Start or stop resimulation based on the current isFocused state
+      unapprovedTransactions.forEach((transactionMeta) => {
```

Comment:
Minor, would it be cleaner to have a `#getActiveUnapprovedTransactions(active: boolean)`?